src/theory/arith/nl/cad/lazard_evaluation.cpp

```diff
@@ -2,6 +2,933 @@
+      const poly::DyadicInterval& di = poly::get_isolating_interval(ran);
+      if (poly::is_point(di))
+      {
+        rational = d_state->convert(poly::get_point(di));
+      }
+      else
+      {
+        Trace("cad::lazard") << "\tis proper ran" << std::endl;
+        polymipo = poly::get_defining_polynomial(ran);
+      }
+    }
+    else
+    {
+      Assert(poly::is_dyadic_rational(val) || poly::is_integer(val)
+             || poly::is_rational(val));
+      if (poly::is_dyadic_rational(val))
+      {
+        rational = d_state->convert(poly::as_dyadic_rational(val));
+      }
+      else if (poly::is_integer(val))
+      {
+        rational = CoCoA::BigRat(d_state->convert(poly::as_integer(val)), 1);
+      }
+      else if (poly::is_rational(val))
+      {
+        rational = d_state->convert(poly::as_rational(val));
+      }
+    }
+
+    if (rational)
+    {
+      d_state->addKRational(var,
+                            CoCoA::RingElem(d_state->d_K.back(), *rational));
+      d_state->d_stats->d_directAssignments++;
+      return;
+    }
+    Trace("cad::lazard") << "Got mipo " << polymipo << std::endl;
+    auto mipo = d_state->convertMiPo(polymipo, var);
+    Trace("cad::lazard") << "Factoring " << mipo << " from "
+                         << CoCoA::owner(mipo) << std::endl;
+    auto factorization = CoCoA::factor(mipo);
+    Trace("cad::lazard") << "-> " << factorization << std::endl;
+    bool used_factor = false;
+    for (const auto& f : factorization.myFactors())
+    {
+      if (d_state->evaluatesToZero(f))
+      {
+        Assert(CoCoA::deg(f) > 0 && CoCoA::NumTerms(f) <= 2);
+        if (CoCoA::deg(f) == 1)
+        {
+          auto rat = -CoCoA::ConstantCoeff(f) / CoCoA::LC(f);
+          Trace("cad::lazard") << "Using linear factor " << f << " -> " << var
+                               << " = " << rat << std::endl;
+          d_state->addKRational(var, rat);
+          d_state->d_stats->d_directAssignments++;
+        }
+        else
+        {
+          Trace("cad::lazard") << "Using nonlinear factor " << f << std::endl;
+          d_state->addK(var, f);
+          d_state->d_stats->d_ranAssignments++;
+        }
+        used_factor = true;
+        break;
+      }
+      else
+      {
+        Trace("cad::lazard") << "Skipping " << f << std::endl;
+      }
+    }
+    Assert(used_factor);
+  }
+  catch (CoCoA::ErrorInfo& e)
+  {
+    e.myOutputSelf(std::cerr);
+    throw;
+  }
+}
+
+void LazardEvaluation::addFreeVariable(const poly::Variable& var)
+{
+  try
+  {
+    d_state->addFreeVariable(var);
+  }
+  catch (CoCoA::ErrorInfo& e)
+  {
+    e.myOutputSelf(std::cerr);
+    throw;
+  }
+}
+
+std::vector<poly::Polynomial> LazardEvaluation::reducePolynomial(
+    const poly::Polynomial& p) const
+{
+  try
+  {
+    return d_state->reduce(p);
+  }
+  catch (CoCoA::ErrorInfo& e)
+  {
+    e.myOutputSelf(std::cerr);
+    throw;
+  }
+  return {p};
+}
+
+/**
```

Comment:
shouldn't this documentation be in the header?

Reply:
I've added some documentation here (in the source file) as well, as I wanted to say a bit about *how* this is implemented. This is irrelevant for the user (and thus not necessary in the header), but should be somewhere.

Reply:
I generally add this kind of high-level operational explanation to the header as well, but I guess this is fine.